File: services/gateway/app/routes/events.ts
import { Router, Request, Response } from 'express';
import axios from 'axios';
import { config } from '../config';
import { proxyRequest } from '../utils/httpProxy';
import { logger } from '../utils/logger';

const router = Router();

// ──────────────────────────────────────────────────────────────────────────────
// GET /v1/events/conversations/:conversationId/stream
// Long-lived Server-Sent Events stream proxied from the orchestrator's Redis
// stream pmos:events:<conversationId>. Emits one SSE message per pipeline event.
// Used by the Decomposition Timeline UI to render a live trace of agent
// actions, decomposition revisions, score evaluations, and sufficiency
// verdicts as the team works.
// ──────────────────────────────────────────────────────────────────────────────
router.get(
  '/events/conversations/:conversationId/stream',
  async (req: Request, res: Response): Promise<void> => {
    const traceId = (req as Request & { id?: string }).id;
    const { conversationId } = req.params;
    const lastEventId =
      (req.headers['last-event-id'] as string | undefined) ??
      (req.query.last_event_id as string | undefined) ??
      '';

    logger.info('events_stream_open', {
      layer: 'router',
      trace_id: traceId,
      conversation_id: conversationId,
      last_event_id: lastEventId,
    });

    // SSE headers: keep the connection open and disable proxy buffering.
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Connection', 'keep-alive');
    res.setHeader('X-Accel-Buffering', 'no');
    res.flushHeaders?.();

    const upstreamUrl =
      `${config.ORCHESTRATOR_URL}/v1/events/conversations/` +
      encodeURIComponent(String(conversationId)) +
      `/stream`;

    const upstreamHeaders: Record<string, string> = {
      accept: 'text/event-stream',
      'x-request-id': traceId ?? '',
    };
    if (lastEventId) {
      upstreamHeaders['last-event-id'] = lastEventId;
    }
    if (req.headers['authorization']) {
      upstreamHeaders['authorization'] = req.headers['authorization'] as string;
    }

    let upstream: { data: NodeJS.ReadableStream; status: number } | null = null;
    try {
      const resp = await axios.get(upstreamUrl, {
        headers: upstreamHeaders,
        responseType: 'stream',
        timeout: 0, // streaming endpoint — no overall timeout
        validateStatus: () => true,
      });
      upstream = resp;

      if (resp.status >= 400) {
        // Drain and end with a JSON error encoded in SSE.
        res.write(`event: error\ndata: {"status":${resp.status}}\n\n`);
        res.end();
        return;
      }

      resp.data.on('data', (chunk: Buffer) => {
        // Pipe upstream SSE bytes verbatim — they're already SSE-formatted
        // by the orchestrator's events route.
        res.write(chunk);
      });

      resp.data.on('end', () => {
        logger.info('events_stream_upstream_ended', {
          layer: 'router',
          trace_id: traceId,
          conversation_id: conversationId,
        });
        res.end();
      });

      resp.data.on('error', (err: Error) => {
        logger.warn('events_stream_upstream_error', {
          layer: 'router',
          trace_id: traceId,
          error: err.message,
        });
        res.end();
      });
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      logger.error('events_stream_failed', {
        layer: 'router',
        trace_id: traceId,
        conversation_id: conversationId,
        error: message,
      });
      res.status(502).end();
      return;
    }

    // If the browser disconnects, close the upstream stream so we don't keep
    // a dead Redis read open.
    req.on('close', () => {
      logger.info('events_stream_client_closed', {
        layer: 'router',
        trace_id: traceId,
        conversation_id: conversationId,
      });
      try {
        // Casting because axios' Readable stream is wrapped.
        const s = upstream?.data as
          | (NodeJS.ReadableStream & { destroy?: () => void })
          | undefined;
        s?.destroy?.();
      } catch {
        // ignore
      }
    });
  },
);

// ──────────────────────────────────────────────────────────────────────────────
// GET /v1/events/conversations/:conversationId/replay
// JSON replay of all events for the conversation, sourced from MySQL
// pipeline_events. Used by the Decomposition Timeline UI for completed runs.
// ──────────────────────────────────────────────────────────────────────────────
router.get(
  '/events/conversations/:conversationId/replay',
  async (req: Request, res: Response): Promise<void> => {
    const { conversationId } = req.params;
    await proxyRequest(req, res, {
      targetUrl:
        `${config.ORCHESTRATOR_URL}/v1/events/conversations/` +
        encodeURIComponent(String(conversationId)) +
        `/replay`,
      timeout: 30_000,
    });
  },
);

export default router;

File: services/orchestrator/app/routes/events.py
"""Pipeline events endpoints — live SSE stream + replay (Phase A).

Surfaces the typed events emitted by ``app.utils.events.EventPublisher`` so the
client's Decomposition Timeline view can render the agentic loop's progress.

Two endpoints:

* ``GET /v1/events/conversations/{conversation_id}/stream`` — Server-Sent Events.
  Tails the Redis stream ``pmos:events:<conversation_id>`` from a configurable
  starting point (``last_event_id`` query param, default ``$`` = only new
  messages). Each SSE message is one event.

* ``GET /v1/events/conversations/{conversation_id}/replay`` — JSON list of all
  events for this conversation, sourced from MySQL ``pipeline_events`` so the
  UI can reconstruct a completed timeline without keeping the stream alive.
"""

from __future__ import annotations

import asyncio
import json
import time
import uuid
from typing import Any, AsyncGenerator, Dict, List, Optional

import mysql.connector
from fastapi import APIRouter, HTTPException, Query, Request
from fastapi.responses import JSONResponse, StreamingResponse

from app.config import settings
from app.utils.events import EventPublisher
from app.utils.logger import logger

router = APIRouter(prefix="/v1/events", tags=["events"])


# ---------------------------------------------------------------------------
# Live SSE stream
# ---------------------------------------------------------------------------


def _format_sse(event: Dict[str, Any], event_id: str) -> str:
    """Format one event as an SSE message. Each line is `field: value`."""
    data = json.dumps(event, default=str)
    # SSE spec: id + event + data + blank line.
    return f"id: {event_id}\nevent: {event.get('kind', 'pipeline.event')}\ndata: {data}\n\n"


async def _tail_stream(
    redis_client: Any,
    stream: str,
    last_id: str,
    block_ms: int = 15000,
) -> AsyncGenerator[bytes, None]:
    """Yield SSE-formatted bytes from the Redis stream.

    Falls through with a heartbeat when ``XREAD`` blocks long enough so the
    HTTP connection stays alive behind proxies that close idle sockets.
    """
    cursor = last_id
    # Initial comment so the EventSource fires `open` and proxies flush headers.
    yield b": connected\n\n"

    while True:
        try:
            results = await redis_client.xread(
                {stream: cursor},
                count=20,
                block=block_ms,
            )
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            logger.warning(
                "events_sse_xread_error",
                layer="router",
                stream=stream,
                error=str(exc)[:300],
            )
            # Brief pause then retry — don't hammer redis if it's flapping.
            await asyncio.sleep(1)
            yield b": redis-retry\n\n"
            continue

        if not results:
            # Heartbeat keeps middleboxes from killing the connection.
            yield b": heartbeat\n\n"
            continue

        for _, messages in results:
            for msg_id, raw in messages:
                cursor = msg_id
                event = _decode_redis_event(raw, msg_id)
                yield _format_sse(event, msg_id).encode("utf-8")
                # If the pipeline signals completion, the client can choose to
                # close the stream. We still keep yielding in case more events
                # arrive (e.g. delayed memory writes).


def _decode_redis_event(raw: Dict[str, Any], msg_id: str) -> Dict[str, Any]:
    """Best-effort decode of Redis stream message back into the publisher's shape."""
    # publish_to_stream stringifies non-string values; payload is JSON-string.
    out: Dict[str, Any] = {"redis_id": msg_id}
    for k, v in (raw.items() if isinstance(raw, dict) else []):
        if k == "payload" and isinstance(v, str):
            try:
                out["payload"] = json.loads(v)
                continue
            except Exception:
                pass
        if k in ("iteration", "round") and isinstance(v, str):
            try:
                out[k] = int(v)
                continue
            except Exception:
                pass
        out[k] = v
    return out


@router.get("/conversations/{conversation_id}/stream")
async def stream_events(
    conversation_id: str,
    request: Request,
    last_event_id: Optional[str] = Query(default=None),
) -> StreamingResponse:
    """Live SSE feed of pipeline events for one conversation.

    The browser ``EventSource`` API automatically reconnects with the
    ``Last-Event-ID`` header on disconnect; we honor it via the
    ``last_event_id`` query param too for explicit control.
    """
    if not conversation_id or len(conversation_id) > 255:
        raise HTTPException(400, "invalid conversation_id")

    redis = request.app.state.redis
    if redis is None or redis._client is None:
        raise HTTPException(503, "redis not connected")

    stream = EventPublisher.stream_key(conversation_id)
    # Prefer header (set automatically by browsers on reconnect), fall back to
    # query param, default to ``$`` which means "only events arriving from now".
    cursor = (
        request.headers.get("last-event-id")
        or last_event_id
        or "$"
    )

    async def gen() -> AsyncGenerator[bytes, None]:
        async for chunk in _tail_stream(redis._client, stream, cursor):
            if await request.is_disconnected():
                logger.info(
                    "events_sse_client_disconnected",
                    layer="router",
                    conversation_id=conversation_id,
                )
                return
            yield chunk

    return StreamingResponse(
        gen(),
        media_type="text/event-stream",
        headers={
            # Disable nginx buffering so events flush immediately.
            "X-Accel-Buffering": "no",
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
        },
    )


# ---------------------------------------------------------------------------
# Replay from MySQL
# ---------------------------------------------------------------------------


def _mysql_cfg() -> Dict[str, Any]:
    return {
        "host": getattr(settings, "mysql_host", "localhost"),
        "port": getattr(settings, "mysql_port", 3306),
        "user": getattr(settings, "mysql_user", "root"),
        "password": getattr(settings, "mysql_password", ""),
        "database": getattr(settings, "mysql_db", "pmos"),
    }


def _replay_sync(conversation_id: str, limit: int) -> List[Dict[str, Any]]:
    """Synchronous fetch of replay rows. Runs in a thread executor."""
    conn = mysql.connector.connect(connection_timeout=5, **_mysql_cfg())
    try:
        cur = conn.cursor(dictionary=True)
        cur.execute(
            """
            SELECT event_id, trace_id, conversation_id, graph_id, node_id,
                   iteration, round, kind, status, payload, ts
            FROM pipeline_events
            WHERE conversation_id = %s
            ORDER BY ts ASC, event_id ASC
            LIMIT %s
            """,
            (conversation_id, int(limit)),
        )
        rows = cur.fetchall()
        cur.close()
        return rows
    finally:
        try:
            conn.close()
        except Exception:
            pass


@router.get("/conversations/{conversation_id}/replay")
async def replay_events(
    conversation_id: str,
    limit: int = Query(default=2000, ge=1, le=10000),
) -> JSONResponse:
    """Reconstruct the timeline for a conversation from MySQL."""
    if not conversation_id or len(conversation_id) > 255:
        raise HTTPException(400, "invalid conversation_id")

    try:
        loop = asyncio.get_event_loop()
        rows = await loop.run_in_executor(None, _replay_sync, conversation_id, limit)
    except Exception as exc:
        logger.error(
            "events_replay_failed",
            layer="router",
            conversation_id=conversation_id,
            error=str(exc)[:300],
        )
        raise HTTPException(500, "replay failed")

    # Decode payload JSON from MySQL string column for client convenience.
    decoded: List[Dict[str, Any]] = []
    for r in rows:
        payload = r.get("payload")
        if isinstance(payload, str):
            try:
                r["payload"] = json.loads(payload)
            except Exception:
                pass
        # Datetime → ISO for JSON.
        ts = r.get("ts")
        if hasattr(ts, "isoformat"):
            r["ts"] = ts.isoformat()
        decoded.append(r)

    return JSONResponse({"conversation_id": conversation_id, "events": decoded})
